File: ruby/lib/runapi/seedream/resources/text_to_image.rb
# frozen_string_literal: true

module RunApi
  module Seedream
    module Resources
      # Seedream text-to-image generation resource.
      # Field requirements vary by model: 4.5/5-Lite/5 Pro require aspect_ratio
      # and output_quality; V4 uses output_resolution and supports seed/output_count.
      class TextToImage
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/seedream/text_to_image"
        RESPONSE_CLASS = Types::TextToImageResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedTextToImageResponse
        PROMPT_MAX_LENGTH = 3000
        V4_PROMPT_MAX_LENGTH = 5000
        PROMPT_MIN_LENGTH_LITE = 3

        def initialize(http)
          @http = http
        end

        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          validate_contract!(CONTRACT["text-to-image"], params)

          model = param(params, :model)

          prompt = param(params, :prompt)
          raise Core::ValidationError, "prompt is required" unless prompt.is_a?(String) && !prompt.empty?
          max_length = Types::LONG_PROMPT_MODELS.include?(model) ? V4_PROMPT_MAX_LENGTH : PROMPT_MAX_LENGTH
          raise Core::ValidationError, "prompt must be at most #{max_length} characters" if prompt.length > max_length
          if Types::MINIMUM_THREE_PROMPT_MODELS.include?(model) && prompt.length < PROMPT_MIN_LENGTH_LITE
            raise Core::ValidationError, "prompt must be between #{PROMPT_MIN_LENGTH_LITE} and #{max_length} characters"
          end

          validate_integer!(params, :seed) if Types::V4_MODELS.include?(model)

          if field_present?(params, :source_image_urls)
            raise Core::ValidationError, "source_image_urls is not supported for #{model}"
          end
        end

        def validate_integer!(params, key)
          value = param(params, key)
          return if value.nil?
          return if value.is_a?(Integer)

          raise Core::ValidationError, "#{key} must be an integer"
        end
      end
    end
  end
end

File: python/src/runapi/seedream/resources/text_to_image.py
"""Seedream text-to-image resource."""

from __future__ import annotations

from typing import Any, Dict

from runapi.core import Resource, ValidationError

from ..contract_gen import CONTRACT
from ..types import (
    LITE_MODELS,
    V4_MODELS,
    CompletedTextToImageResponse,
    TextToImageResponse,
)


class TextToImage(Resource):
    """Generate images from text prompts with Seedream models."""

    ENDPOINT = "/api/v1/seedream/text_to_image"

    RESPONSE_CLASS = TextToImageResponse
    COMPLETED_RESPONSE_CLASS = CompletedTextToImageResponse

    PROMPT_MAX_LENGTH = 3000
    V4_PROMPT_MAX_LENGTH = 5000
    PROMPT_MIN_LENGTH_LITE = 3

    def run(self, **params: Any) -> Any:
        """Create a text-to-image task and poll until it completes.

        Args:
            **params: text-to-image parameters (model, ...).

        Returns:
            The completed (narrowed) text-to-image response.
        """
        task = self.create(**params)
        return self._poll_until_complete(lambda: self.get(task.id))

    def create(self, **params: Any) -> Any:
        """Create a text-to-image task and return immediately with an id.

        Args:
            **params: text-to-image parameters (model, ...).

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted)

    def get(self, id: str) -> Any:
        """Fetch the current status of a text-to-image task.

        Args:
            id: The task id returned by ``create``.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}")

    def _validate_params(self, params: Dict[str, Any]) -> None:
        self._validate_contract(CONTRACT["text-to-image"], params)

        model = params.get("model")

        prompt = params.get("prompt")
        if not (isinstance(prompt, str) and prompt):
            raise ValidationError("prompt is required")
        max_length = self.V4_PROMPT_MAX_LENGTH if model in V4_MODELS else self.PROMPT_MAX_LENGTH
        if len(prompt) > max_length:
            raise ValidationError(f"prompt must be at most {max_length} characters")
        if model in LITE_MODELS and len(prompt) < self.PROMPT_MIN_LENGTH_LITE:
            raise ValidationError(
                f"prompt must be between {self.PROMPT_MIN_LENGTH_LITE} and {self.PROMPT_MAX_LENGTH} characters"
            )

        if model in V4_MODELS:
            self._validate_integer(params, "seed")

        if self._field_present(params, "source_image_urls"):
            raise ValidationError(f"source_image_urls is not supported for {model}")

    @staticmethod
    def _field_present(params: Dict[str, Any], key: str) -> bool:
        value = params.get(key)
        if value is None:
            return False
        if hasattr(value, "__len__"):
            return len(value) > 0
        return True

    @staticmethod
    def _validate_integer(params: Dict[str, Any], key: str) -> None:
        value = params.get(key)
        if value is None:
            return
        if isinstance(value, int) and not isinstance(value, bool):
            return
        raise ValidationError(f"{key} must be an integer")
